PsfShimMonitor/KernelTrace.cs

```diff
@@ -0,0 +1,1356 @@
+                                    }
+                                }
+                                catch {  /* event thrown if key exists, which should not happen here */ }
+                            }
+                        }
+                        _TKCBs.Clear();
+                    }
+                }
+                lock (_TKernelEventListsLock) //_TKCBsListLock)
+                {
+                    if (_TKernelEventListItems.Count > 0)
+                    {
+                        foreach (EventItem ei in _TKernelEventListItems)
+                        {
+                            AppplyFilterToEventItem(ei);
+                            if (IsPaused)
+                                ei.IsPauseHidden = true;
+                            ApplyPastKCBsToRegistryEvent(ei);
+                            _ModelEventItems.Add(ei);
+                        }
+                        _TKernelEventListItems.Clear();
+
+                        UpdateFilteredViewList();
+                    }
+                    else
+                        EventsGrid.Items.Refresh();
+                }
+            }
+            catch
+            {
+                ;
+            }
+        }
+
+        private void KernelTrace_RunWorkerCompleted(object sender, RunWorkerCompletedEventArgs e)
+        {
+            Status.Text = "NonKernel";
+        }
+
+        private void ApplyKCBtoPastRegistryEvents(UInt64 k, string s)
+        {
+            foreach (EventItem ei in _ModelEventItems)
+            {
+                if (ei.EventSource.StartsWith("Registry/"))
+                {
+                    if (ei.Inputs.Contains("KeyName=\n") &&
+                        !ei.Inputs.Contains(")\nKeyName="))
+                    {
+                        if (ei.Inputs.Contains("Key=      \t" + k.ToString() + "\n"))
+                        {
+                            ei.Inputs.Replace("\nKeyName=", " (" + s + ")\nKeyName=");
+                        }
+                    }
+                }
+            }
+        }
+        private void ApplyPastKCBsToRegistryEvent(EventItem ei)
+        {
+            if (ei.Event.StartsWith("Registry/"))
+            {
+                if (ei.Inputs.Contains("KeyName=\n") &&
+                    !ei.Inputs.Contains(")\nKeyName="))
+                {
+                    string matchh = ei.Inputs.Substring(10, ei.Inputs.IndexOf('\n') - 8);
+                    foreach (UInt64 key in _KCBs.Keys)
+                    {
+                        if (key.ToString().Equals(matchh))
+                        {
+                            try
+                            {
+                                ei.Inputs.Replace("\nKeyName=", " (" + _KCBs[key] + ")\nKeyName=");
+                            }
+                            catch { }
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        int TSM_ProcID;
```

Comment:
TSM_ProcID is being used only in EnableKernelTrace.  Can you please move it into the method?

Reply:
It is used accessed by the delegate which continues to live after the scope of this method completes, thus needs to be of class scope..

Reply:
Right.  I am so sorry.  I haven't worked with delegates in a while.